## Filtering and annotating somatic variant calls

This folder contains scripts for filtering somatic variant calls (as maf files) generated by intersecting Mutect2 and Strelka2 outputs using VariantFilter as described in the methods. Code used to call these algorithms on raw sequencing data is not provided here- to reproduce the unfiltered calls, please follow the directions in the Methods section of our manuscript. Also adds copy number calls at variant loci to the mafs for ppVAF estimation downstream. Please run these notebooks in the order they are presented (`filter_blacklist_shared.ipynb` and then `add_CN_mafs.ipynb`) to reproduce these steps. Note that the output mafs from the notebooks in this directory are not the final filtered mafs, since they do not include ppVAF estimates. Use the scripts in `analysis/ppVAF_calculations` to then generate the final mafs using these filtered outputs.

### Description of scripts

`filter_blacklist_shared.ipynb`

Reads VariantFilter-created somatic mutation tables that have been converted to mafs using vcf2maf and filters out those in telomeres and other blacklisted regions of the genome as those with other properties associated with technical artifacts (see Methods). Also removes somatic variants shared between multiple lesions in the HTAN dataset, except those whitelisted as driver mutations that could have arisen independently multiple times. Saves filtered and annotated mafs for all datasets processed. 

Requires:
* unfiltered HTAN WGS and WES mafs (found in Zenodo dataset as `HTAN_WGS_unfiltered.maf` and `HTAN_WES_unfiltered.maf`)
* unfiltered SCORT and PUTH mafs (not provided, would have to generate these yourself following the preprocessing pipeline described in the Methods)
* hg38 blacklisted regions from UCSC (in this repo at `data/resource/hg38.UCSC.centromere.telomere.encode.bed`)
* COADREAD driver gene list (in this repo at `data/resource/PanCanDrivers_COADREAD_Cell2018.csv`)

Generates:
* filtered HTAN WGS and WES mafs **without** copy number information (temp filenames used are `*_filtered.maf`)
* filtered PUTH and SCORT mafs if those external datasets were used as inputs

`add_CN_mafs.ipynb`

Adds major and minor copy number allele information from the filtered FACETS copy number calls (generated in the scripts found in `analysis/copy_number`).

Requires:
* filtered copy number calls with filtered gaps filled in by merging adjacent segments (HTAN calls in this repo at `data/copy_number/*CN_filtered_merged.tsv`)
* filtered HTAN WGS and WES mafs **without** copy number information (temp filenames used are `*_filtered.maf`)
* filtered PUTH and SCORT mafs if filtered mafs are available for these external datasets

Generates:
* filtered mafs with copy number information but without ppVAF estimates (temp filenames used are `*_filtered_CNs.maf`)

