## Data repository for bulk HTAN genomic sequencing

This directory contains data and resource files generated by and/or used for analysis of the bulk WES and WGS data.  

Larger files like mafs can be found in our Zenodo repository. Most of the intermediate data files are generated by other scripts in this repo (see `bulk_analyses/analysis` for code that generates them) with a few notable exceptions:

`telseq_estimates_wgs_ff.tsv`: Generated by running TelSeq, see Methods

`HTAN_WGS_facets_sequenza_purities.csv` and `HTAN_WES_facets_sequenza_purities.csv`: Generated from FACETS and Sequenza outputs, see Methods

`./scATACseq_annotations/*celltype_atac.tsv`: published in [Becker et al 2022](https://www.nature.com/articles/s41588-022-01088-x) and with data files downloaded from [their Google Drive directory](https://drive.google.com/drive/folders/1Kl3SSbQyYQWIzl1ZW_9AfkFJStr7sY8A).

`./coverage/*`: Median coverage tables generated by running [mosdepth](https://github.com/brentp/mosdepth), see Methods

This directory also includes some metadata files describing the sample types, etc in the HTAN (`Table S2`) and external (`./metadata`) cohorts.
