## Filtering and processing copy number calls

This folder contains scripts for filtering copy number calls generated by [FACETS](https://github.com/mskcc/facets-suite) and estimating the fraction genome altered (FGA) using the filtered copy number calls. Code used to call FACETS on our raw sequencing data is not provided here- to reproduce the unfiltered calls, please follow the directions in the Methods section of our manuscript. Please run these notebooks in the order they are presented (`filter_FACETS_calls.ipynb` and then `compute_FGA.ipynb`) to reproduce these steps.

### Description of scripts

`filter_FACETS_calls.ipynb`

Reads raw unfiltered FACETS copy number calls (see input table formatting in `data/copy_number/HTAN_WGS_CN_unfiltered.tsv`) and filters out segments with high overlap with telomeres and other blacklisted regions of the genome as those with other properties associated with technical artifacts (see Methods). Saves filtered copy number calls for all datasets processed. Note that this notebook generates a number of temporary files that are not distributed with this repo, you will have to specify a location to save these temporary files.

Requires:
* unfiltered HTAN WGS and WES FACETS copy number calls (included in this repo as `data/copy_number/HTAN_WGS_CN_unfiltered.tsv` and `data/copy_number/HTAN_WES_CN_unfiltered.tsv`)
* unfiltered FACETS calls from external datasets (PUTH and SCORT) if desired (not included in our data deposition, you must request the raw data from these studies and generate these yourself, they should be in the same format as the HTAN examples published here)
* FACETS whole genome doubling calls (in this repo at `data/copy_number/genome_doubling/HTAN_WGS_doubled.tsv/` and `data/copy_number/genome_doubling/HTAN_WES_doubled.tsv/` for the HTAN datasets; would have to generate them yourself for PUTH and SCORT)
* table of hg38 gene coordinates (provided here as `data/resource/hg38_gene_locs.bed` and generated from the [GENCODE human basic annotation](https://www.gencodegenes.org/human/) with the code in the Appendix of this notebook)
* hg38 blacklisted regions from UCSC (in this repo at `data/resource/hg38.UCSC.centromere.telomere.encode.bed`)

Generates:
* filtered copy number calls with gaps for regions filtered out (provided in this repo at `data/copy_number/*CN_filtered.tsv`) and with gaps filled in by merging adjacent segments (`data/copy_number/*CN_filtered_merged.tsv`). Merged calls were used to identify the major and minor allele copy numbers at somatic mutation loci and to estimate the FGAs.
* gene-level copy number calls used for oncoplots (provided for the HTAN datasets in this repo at `data/copy_number/gene_CN_calls/HTAN_WGS_gene_CNs.tsv` and `data/copy_number/gene_CN_calls/HTAN_WES_gene_CNs.tsv`)

`compute_FGA.ipynb`

Uses the filtered copy number calls to estimate and save the fraction genome altered (FGA) for each sample.

Requires:
* filtered copy number calls with filtered gaps filled in by merging adjacent segments (HTAN calls in this repo at `data/copy_number/*CN_filtered_merged.tsv`)
* hg38 blacklisted regions from UCSC (in this repo at `data/resource/hg38.UCSC.centromere.telomere.encode.bed`)

Generates:
* fraction genome altered tables (HTAN calls in this repo at `data/fraction_genome_altered/*_fga.tsv`)

`CN_utils.py`

Contains helper functions used in `filter_FACETS_calls.ipynb` and `compute_FGA.ipynb`. No need to run this script on its own, but take a look inside for dependencies if running the other scripts.
